feature: VRMAnimation

0b5vr opened the pull request:
### Description

Add a new module `@pixiv/three-vrm-animation` which supports the [VRMC_vrm_animation](https://github.com/vrm-c/vrm-specification/tree/master/specification/VRMC_vrm_animation-1.0) extension (which is still a draft).

The VRMAnimation code already existed for projects like "ChatVRM" or "bvh2vrma", but we updated several interfaces and doc comments for the OSS purpose.

There are two examples in the three-vrm-animation package, `loader-plugin.html` and `dnd.html`; the former is slightly easier to understand, while the latter is helpful in checking its behaviors.

### APIs

- `@pixiv/types-vrmc-vrm-animation-1.0`
    - The type definition of the `VRMC_vrm_animation` extension.
    - `VRMCVRMAnimation` represents the extension root of `VRMC_vrm_animation`.
- `@pixiv/three-vrm-animation`
    - A module to load VRM Animation.
    - `VRMAnimationLoaderPlugin`, a plugin of `THREE.GLTFLoader`, loads a `.vrma` binary and creates `VRMAnimation`s.
    - `VRMAnimation` can be converted to `THREE.AnimationClip` using `createVRMAnimationClip`. Since the animation clip depends on the target model's structure, `VRM` must also be specified.

### Points need review

- [x] APIs. Do you think it looks good?
    - What do you think about the decision that separated `createVRMAnimationClip` from methods of `VRMAnimation`? I thought it would be safer to make it loose-coupled
- [x] Examples. Do you think it's good enough to convey how it should be used?
- [x] Is it able to build on your end?
- [x] Repository structure. Is it okay to make three-vrm-animation join this repository? Considering the maintenance cost, I think having it in this repository is better than making a separate repo.

0b5vr committed:
feature: add a new type package, `@pixiv/types-vrmc-vrm-animation-1.0`

It has a type definition for `VRMC_vrm_animation`, the extension is still draft

The spec: https://github.com/vrm-c/vrm-specification/tree/0861a66eb2f2b76835322d775678047d616536b3/specification/VRMC_vrm_animation-1.0

0b5vr committed:
feature: Implement VRMAnimation

The VRMAnimation code already existed for projects like "ChatVRM" or "bvh2vrma"
We updated several interfaces and doc comments for the OSS purpose

It uses the type package `@pixiv/types-vrmc-vrm-animation-1.0` we've created in the previous commit

There are two examples, `loader-plugin.html` and `dnd.html`; the former is slightly easier to understand while the latter is useful to check its behaviors

0b5vr committed:
chore: Fix lint issues

Top-level await is not supported natively in eslint v7 :(
https://eslint.vuejs.org/user-guide/#using-eslint-v8-x

also add missing `.editorconfig` and `.eslintrc`

0b5vr committed:
chore: suppress a eslint warning

@typescript-eslint/no-unused-vars doesn't consider doc comments

0b5vr committed:
fix: fix build errors of three-vrm-animation

- test job should not call jest since there is no test
- it's dependant of `@pixiv/three-vrm-core`

0b5vr committed:
fix: `VRMHumanBoneParentMap` of `three-vrm-core` should be included in `three-vrm-animation` build

add `@rollup/plugin-node-resolve` to rollup config of `three-vrm-animation`

ke456-png commented on packages/three-vrm-animation/src/createVRMAnimationClip.ts:
```diff
@@ -0,0 +1,103 @@
+    tracks.push(...createExpressionTracks(vrmAnimation, vrm.expressionManager));
+  }
+
+  if (vrm.lookAt != null) {
+    const lookAtTarget = vrm.lookAt.target;
+
+    if (lookAtTarget == null) {
+      console.warn('`vrm.lookAt.target` of the given VRM is not defined. Skipping lookAt animation');
+    } else {
+      const track = createLookAtTrack(vrmAnimation, `${lookAtTarget.name}.position`);
+
+      if (track != null) {
+        tracks.push(track);
+      }
+    }
+  }
```
無名のObjectをLookAtTargetに指定してしまった場合、アニメーションがrootに適用されてしまう点が気になります。

```js
const lookAtTarget = new THREE.Object3D();
// lookAtTarget.name = 'lookAtTarget';
vrm.scene.add( lookAtTarget );
vrm.lookAt.target = lookAtTarget;
```

Objectにデフォルトでは名前がないので引っかかる場合が多いと考えています。
`lookAtTarget.name == null`でも判定するのはどうでしょうか？

ke456-png replied:
LookAt.targetにはcameraを指定することも多いと考えられますが、その場合vrm.sceneにcameraを追加すると動いてしまいますね……。

ただ、あまり`vrm.scene.add( camera );`はすることはないと思うので対応無しでも良いとは思います。

0b5vr replied:
> Objectにデフォルトでは名前がないので引っかかる場合が多いと考えています。
> lookAtTarget.name == nullでも判定するのはどうでしょうか？

なるほどね～。ご指摘ありがとうございます。
名前がない場合に動かなくするよりかは、warningを出すくらいが良さそうかな？と思います。
ついでに、 `lookAtTarget` とかに自動リネームしても良いかもなと思います。

ke456-png replied:
> 名前がない場合に動かなくするよりかは、warningを出すくらいが良さそうかな？と思います。
> ついでに、 lookAtTarget とかに自動リネームしても良いかもなと思います。

👍

0b5vr replied:
Object3Dのお名前はコンストラクタで `''` が設定されていることが保証されていそうなので、コードの読みやすさを考えて `lookAtTarget.name === ''` でいきます

https://github.com/mrdoob/three.js/blob/r157/src/core/Object3D.js#L40

ke456-png replied:
> Object3Dのお名前はコンストラクタで '' が設定されている

そうなんですね！方針に同意します。

ke456-png replied:
Resolved.

https://github.com/pixiv/three-vrm/pull/1290/commits/0ba883279bfefa9b19f66ed440e5743569a2d11e

ke456-png commented on packages/three-vrm-animation/src/createVRMAnimationClip.ts:
```diff
@@ -0,0 +1,103 @@
+ * @param vrm A {@link VRMCore}.
+ * @returns An AnimationClip
+ */
+export function createVRMAnimationClip(vrmAnimation: VRMAnimation, vrm: VRMCore): THREE.AnimationClip {
```
expression, humanoid, lookAtのclipを分けて扱いたいシチュエーションに対応したいです。
例) モーションの表情だけ別で制御したい

- optionで特定のtrackを無効化出来るようにしてcreateVRMAnimationClipを複数回叩いてもらう
- trackやclipを分割して返せるようにする

等の解決策をイメージしています。

0b5vr replied:
今、 `createVRMAnimationClip()` で `createHumanoidTracks` 等を直接叩いて、そのまま `AnimationClip` を返していますが、上に定義されている `createHumanoidTracks()` とかもexposeしましょうか。
その上で、 `track.userData` にどこのボーンに対応しているかを……　ああ、 `THREE.KeyframeTrack.userData` は定義されていない……

0b5vr replied:
`createHumanoidTracks()` の返り値を `Record<VRMHumanBoneName, THREE.VectorKeyframeTrack>` にするのは良いかなと思いました。 `createExpressionTracks()` も同様。

あと、これらをexposeするとなると、 `createLookAtTrack` だけtrack名を渡す仕組みというのはちょっとinconsistentかなあと思いました。代わりに `VRMLookAt` を渡すのがよいのかなと思っておりますが、いかがでしょう？

ke456-png replied:
> ` create***Tracks()`をexposeする。

同意します！

> その上で、 track.userData にどこのボーンに対応しているかを……

とりあえず、ボーンごとのマスクがしたい場合はtrack.nameを参照してもらうのはどうでしょうか？

> createLookAtTrack だけtrack名を渡す仕組みというのはちょっと
> ...
> 代わりに VRMLookAt を渡すのがよいのかなと思っておりますが、いかがでしょう？

VRMLookAtが良いと思います。
VRMLookAtかLookAtTargetの2択で考えてみると、後者の型がただのObject3dで情報が無いので。

ke456-png replied:
expression, humanoid, lookAtの個別importに関してはcreate***Tracks()をexposeするという方針で良いと思います。

> createHumanoidTracks() の返り値を Record<VRMHumanBoneName, THREE.VectorKeyframeTrack> にするのは良いかなと思いました。 createExpressionTracks() も同様。

これはさらに細かく個別で制御したい場合ですよね、少し過剰かもしれないと思ったのですが、
プロシージャルな表情(例：まばたき)と衝突しないように制御したいとかは普通にありそうですね 😵‍💫

0b5vr replied:
いわゆるモーションのマスクの需要が普通にありそうだなと思って。
Hipsから下・Hipsから上で適用するモーション変えるとか

ke456-png replied:
複数のモーションを組む合わせるのは確かにありますね。

そうなると`VRMHumanBoneName`をkeyにしたRecordにするのは良いですね！

Hipは2つトラックがある点やcustom expressionの扱いが課題だと思います

0b5vr replied:
確かに…… > Hipsは2つトラックがある

Humanoidのほうは、
```ts
{
  position: Partial<Record<'hips', THREE.VectorKeyframeTrack>>;
  rotation: Partial<Record<VRMHumanBoneName, THREE.QuaternionKeyframeTrack>>;
}
```

Expressionsのほうは、
```ts
{
  preset: Partial<Record<VRMExpressionPresetName, THREE.NumberKeyframeTrack>>;
  custom: Record<string, THREE.NumberKeyframeTrack>;
}
```
みたいなノリがまるそうですが、いかがでしょうか？

0b5vr replied:
いや、 `VRMAnimation` のメンバのほうが `Map` なので、ここの返り値もMapのほうが良いか

0b5vr replied:
さらに言うとExpressionsのほうも `VRMAnimation` のほうはpresetとcustomを分けてないですね。これは `VRMAnimation` 側も併せて検討が必要か

0b5vr replied:
`VRMAnimation` , `position` のトラックが `hips` 以外も読めるようになってるので、別PRで以下みたいな処理をLoaderに加えます:

```ts
          if (boneName !== 'hips') {
            console.warn(`The loading animation contains a translation track for ${boneName}, which is not permitted in the VRMC_vrm_animation spec. skipping the track`);
          } else {
```

0b5vr replied:
https://github.com/pixiv/three-vrm/pull/1290/commits/86ea3157ba88ac79cb3e2661e6fa59c2a2c63cfd でexposeしました。
exposeするのに名前空間を調整したところ、お名前がだいぶ長くなっちゃいました 😅

0b5vr replied:
> `VRMAnimation` , `position` のトラックが `hips` 以外も読めるようになってるので、別PRで以下みたいな処理をLoaderに加えます:

別PRって、このPR自体がVRMAnimationを追加しているのか……

追加で作業して、わかりやすいように別Conversationを建てます。

0b5vr committed:
change: createVRMAnimationClip, set a name to lookAtTarget automatically if the target is not named

This emit a warning also

See: https://github.com/pixiv/three-vrm/pull/1290#discussion_r1364978190

0b5vr committed:
change: VRMAnimation, expose createVRMAnimationHumanoidTracks, createVRMAnimationExpressionTracks, createVRMAnimationLookAtTrack

- change the name of these functions to tolerate on the exposal of functions
- Change return type of these functions to make end developers able to filter out unnecessary tracks

See: https://github.com/pixiv/three-vrm/pull/1290#discussion_r1369842746

0b5vr committed:
change: VRMAnimation and createVRMAnimationExpressionTracks, separate preset and custom expressions

See: https://github.com/pixiv/three-vrm/pull/1290#discussion_r1369897162

0b5vr committed:
change: VRMAnimation is now unable to have translation tracks for bones other than hips

See: https://github.com/pixiv/three-vrm/pull/1290#discussion_r1369911275

0b5vr committed:
change: createVRMAnimationHumanoidTracks, rename `.position` -> `.translation` for consistency with VRMAnimation

0b5vr commented on packages/three-vrm-animation/src/VRMAnimation.ts:
```diff
@@ -0,0 +1,40 @@
+    rotation: Map<VRMHumanBoneName, THREE.QuaternionKeyframeTrack>;
+  };
+  public expressionTracks: {
+    preset: Map<VRMExpressionPresetName, THREE.NumberKeyframeTrack>;
+    custom: Map<string, THREE.NumberKeyframeTrack>;
```
presetとcustomを分けた

https://github.com/pixiv/three-vrm/pull/1290/commits/a2e9b7ef62e75b7e50be215ed235055076bc4b64

0b5vr commented on packages/three-vrm-animation/src/VRMAnimation.ts:
```diff
@@ -0,0 +1,40 @@
+  public restHipsPosition: THREE.Vector3;
+
+  public humanoidTracks: {
+    translation: Map<'hips', THREE.VectorKeyframeTrack>;
```
`translation` のkeyを `'hips'` のみに制限した

https://github.com/pixiv/three-vrm/pull/1290/commits/619fbb72f441148f3aaa6f824740c031ba069210

0b5vr commented on packages/three-vrm-animation/src/createVRMAnimationClip.ts:
```diff
@@ -0,0 +1,141 @@
+  humanoid: VRMHumanoid,
+  metaVersion: '0' | '1',
+): {
+  translation: Map<'hips', THREE.VectorKeyframeTrack>;
```
名前を `VRMAnimation` とconsistentに。 `rotation` → `translation`

https://github.com/pixiv/three-vrm/pull/1290/commits/4c746b997e7e92fe664a473ec8e59b35a536c56b

0b5vr committed:
fix (vrm-animation): fix the loader behavior when humanoid bones are not defined

There is a case that humanoid does not exist in the animation file when the file only have expressions for example

0b5vr committed:
change: VRMAnimation LookAt is now local space quaternion instead of model space translation

See: https://github.com/vrm-c/vrm-specification/pull/456

transfer quaternion to VRMLookAt yaw-pitch using VRMLookAtQuaternionProxy (does it look good?)

ke456-png commented on packages/three-vrm-animation/src/VRMLookAtQuaternionProxy.ts:
```diff
@@ -0,0 +1,39 @@
+
+    this.type = 'VRMLookAtQuaternionProxy';
+
+    // See: https://github.com/mrdoob/three.js/blob/r158/src/core/Object3D.js#L65
+    const prevRotationOnChangeCallback = this.rotation._onChangeCallback;
+    this.rotation._onChange(() => {
+      prevRotationOnChangeCallback();
+      this._applyToLookAt();
+    });
+
+    const prevQuaternionOnChangeCallback = this.quaternion._onChangeCallback;
+    this.quaternion._onChange(() => {
+      prevQuaternionOnChangeCallback();
+      this._applyToLookAt();
+    });
+  }
```
👍

ke456-png commented on packages/three-vrm-animation/src/createVRMAnimationClip.ts:
```diff
@@ -117,23 +118,31 @@ export function createVRMAnimationClip(vrmAnimation: VRMAnimation, vrm: VRMCore)
-      if (track != null) {
-        tracks.push(track);
-      }
+    // search VRMLookAtQuaternionProxy
+    let proxy = vrm.scene.children.find((obj) => obj instanceof VRMLookAtQuaternionProxy);
+
+    if (proxy == null) {
+      // if not found, create a new one
+      console.warn(
+        'createVRMAnimationClip: VRMLookAtQuaternionProxy is not found. Creating a new one automatically. To suppress this warning, create a VRMLookAtQuaternionProxy manually',
+      );
+
+      proxy = new VRMLookAtQuaternionProxy(vrm.lookAt);
+      proxy.name = 'VRMLookAtQuaternionProxy';
+      vrm.scene.add(proxy);
+    } else if (proxy.name == null) {
+      // if found but name is not set, set the name automatically
+      console.warn(
+        'createVRMAnimationClip: VRMLookAtQuaternionProxy is found but its name is not set. Setting the name automatically. To suppress this warning, set the name manually',
+      );
+
+      proxy.name = 'VRMLookAtQuaternionProxy';
+    }
```
👍 
Resolved https://github.com/pixiv/three-vrm/pull/1290#pullrequestreview-1686825559

ke456-png commented on packages/three-vrm-animation/examples/loader-plugin.html:
```diff
@@ -98,11 +98,10 @@
-				vrm.scene.add( lookAtTarget );
-				vrm.lookAt.target = lookAtTarget;
+				// Add look at quaternion proxy to the VRM; which is needed to play the look at animation
+				const lookAtQuatProxy = new VRMLookAtQuaternionProxy( vrm.lookAt );
+				lookAtQuatProxy.name = 'lookAtQuaternionProxy';
+				vrm.scene.add( lookAtQuatProxy );
```
[memo]lookAtQuatProxyの親はvrm.scene

座標系はモデルのRoot座標系

ke456-png reviewed (approved):
NICE! 👍

0b5vr commented:
will release the `@pixiv/three-vrm-animation` as beta version
